### PR TITLE
Resolve issue #981

### PR DESCRIPTION
**Pull request by boliza**
Resolve issue #981

### CODE REVIEW

**Comment by dalaro** [comment]
Why is this necessary?  At first I assumed this was on the read path -- an edge case where an element comes back as it is expiring.  However, that doesn't seem to be the case.

This PR modifies one of two `ConversionHelper.getTTLSeconds` method overloads.  In practice it modifies both, since the unmodified overload just does some trivial parameter conversion and then calls the modified overload.

The two `getTTLSeconds` methods collectively have two call sites (not counting one overload calling the other):
- Setting the TTL on a `KCVSLogManager`'s store.  I don't see why we need 0 TTL here.  Isn't that the same as laboriously writing the log to `/dev/null`?
- Turning the `ttl` and `unit` parameters to `ManagementSystem.setTTL(type, ttl, unit)` into seconds.  This controls the TTL at the schema level.  My concern here is in the same vein as my concern for `KCVSLogManager`; what is the use of a type that dies as it is written?

Can you help me understand why this change is necessary?  Also, can you confirm that the #92 issue reference in the commit message was a typo?  I'm assuming you meant #982 or #981.

**Comment by boliza** [comment]
> Can you help me understand why this change is necessary? Also, can you confirm that the #92 

this is my mistake ,sorry! It is means 981

**Comment by boliza** [comment]
@dalaro You are right.It's not necessary, but sometimes when want store some data with ttl before forever.The only way we should do is ,add new property

**Comment by dalaro** [comment]
> It's not necessary, but sometimes when want store some data with ttl before forever.The only way we should do is ,add new property

I'm struggling to understand.  Can you rephrase this?  I still don't understand why zero-ttl (instantly deleted) elements are useful.

**Comment by boliza** [comment]
I am not good at English.When I design my graph , I think some properties should have ttl,but after a period of time ,because of the change of business（requirement）, those properties with ttl must be store without ttl.This just my requirement.You can think about it.

Thanks for you job!!!

**Comment by dalaro** [comment]
Thanks, that's very helpful, I think I finally understand.

Titan's API is asymmetrical here:
- Javadoc for `TitanManagement.getTTL` says 0 TTL means live forever:
  
  ```
  Retrieves the time-to-live for the given {@link TitanSchemaType} as a {@link Duration}.
  If none has been explicitly defined, a zero-length {@link Duration} is returned.
  ```
- Javadoc for `TitanManagement.setTTL` is silent about what 0 TTL means.  The implementation treats 0 as invalid in practice and throws this exception.

This is a deficiency in Titan: `setTTL` and `getTTL` should interpret all values the same way, as you (reasonably) expected with 0.  I think I would actually prefer to make -1 mean "live forever" and disallow 0 (since 0 is more ambiguous than -1).

@boliza I have a separate concern about your use case.  Even if we make `TitanManagement.setTTL(<some-type>, <live-forever>)` possible, it may not be possible to update TTL on existing data.  Old data of `<some-type>` could expire even after changing the TTL.  New elements written after committing `TitanManagement.setTTL(<some-type>, <live-forever>)` should live forever.  Does that make sense?

**Comment by boliza** [comment]
You are right,can't change the old data,just affect the new elements.

**Comment by joshsh** [comment]
You are probably aware, but Titan inherits 0 as the default, infinite TTL from Cassandra. See [CQL v3](http://cassandra.apache.org/doc/cql3/CQL.html).  It might be wise to maintain that agreement, even to the extent of ensuring that negative values are also treated as infinite.  setTTL should indeed accept 0, particularly if TTL is mutable.

**Comment by dalaro** [comment]
@joshsh Indeed, I abandoned the -1 idea once I got into the implementation.  It's not only at odds with Cassandra convention, but also a poor fit for Titan internals.  StandardDuration only accepts nonnegatives by design, and 0 is already hardcoded in several places as a "live forever" default value.  I recently replied about that on related issue #981.  Sorry to jump back and forth between here and there.

**Comment by dalaro** [comment]
I think this has been superseded by #1069.  Please reopen #981 for comment if I missed something.
